liblouis/compileTranslationTable.c

```diff
@@ -3763,43 +3745,136 @@ compileRule(FileInfo *nested, TranslationTableOffset *newRuleOffset,
+
+			ok = 1;
+			if (!getToken(nested, &token, "attribute name", &lastToken)) {
+				compileError(nested, "Expected %s", "attribute name");
+				ok = 0;
+				break;
 			}
-			if (getToken(nested, &token, "character class name", &lastToken)) {
-				class = findCharacterClass(&token, *table);
-				if (!class)
-					// no class with that name: create one
-					class = addCharacterClass(
-							nested, &token.chars[0], token.length, *table);
-				if (class) {
-					// there is a class with that name or a new class was successfully
-					// created
-					if (getCharacters(nested, &characters, &lastToken)) {
-						int index;
-						for (index = 0; index < characters.length; ++index) {
-							TranslationTableRule *defRule;
-							// get the character from the table and add the new class to
-							// its attributes if the character is not defined yet, define
-							// it
-							TranslationTableCharacter *character =
-									putChar(nested, characters.chars[index], table);
-							character->attributes |= class->attribute;
-							// also add the attribute to the associated dots (if any)
-							if (character->definitionRule) {
-								defRule = (TranslationTableRule *)&(*table)
-												  ->ruleArea[character->definitionRule];
-								if (defRule->dotslen == 1) {
-									character =
-											getDots(defRule->charsdots[defRule->charslen],
-													*table);
-									if (character)
-										character->attributes |= class->attribute;
+
+			if (!(*table)->characterClasses && !allocateCharacterClasses(*table)) {
+				ok = 0;
+				break;
+			}
+
+			TranslationTableCharacterAttributes attribute = 0;
+			{
+				int attrNumber = -1;
+				switch (token.chars[0]) {
+				case '0':
+				case '1':
+				case '2':
+				case '3':
+				case '4':
+				case '5':
+				case '6':
+				case '7':
+				case '8':
+				case '9':
+					attrNumber = token.chars[0] - '0';
+					break;
+				}
+				if (attrNumber >= 0) {
+					if (opcode == CTO_Class) {
+						compileError(nested,
+								"Invalid class name: may not contain digits, use "
+								"attribute instead of class");
+						ok = 0;
+						break;
+					} else if (token.length > 1 || attrNumber > 7) {
+						compileError(nested,
+								"Invalid attribute name: must be a digit between 0 and 7 "
+								"or a word containing only letters and digits");
+						ok = 0;
+						break;
+					}
+					if (!(*table)->numberedAttributes[attrNumber])
+						// attribute not used before yet: assign it a value
+						(*table)->numberedAttributes[attrNumber] =
+								getNextNumberedAttribute(*table);
+					attribute = (*table)->numberedAttributes[attrNumber];
+				} else {
+					const CharacterClass *namedAttr = findCharacterClass(&token, *table);
+					if (!namedAttr) {
+						// no class with that name: create one
+						for (i = 0; i < token.length; i++) {
+							if (!((token.chars[i] >= 'a' && token.chars[i] <= 'z') ||
+										(token.chars[i] >= 'A' &&
+												token.chars[i] <= 'Z'))) {
```

Comment:
Hm, this seems odd. Shouldn't we check for digits as well here? At least the warning implies that.

Reply:
But there should not be digits in the names.